main/src/library/MutHeap.flix

```diff
@@ -0,0 +1,240 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pub enum MutHeap[a: Type, r: Region] {
+    case MutHeap(Region[r], Ref[Array[a, r], r], Ref[Int32, r])
+}
+
+mod MutHeap {
+
+    ///
+    /// Returns an empty MutHeap
+    ///
+    pub def new(rc: Region[r]): MutHeap[a, r] \ r =
+        MutHeap(rc, ref Array.new(rc, 8) @ rc, ref 0 @ rc)
+
+    ///
+    /// Returns an empty MutHeap with a reserved amount of space
+    ///
+    pub def newWithSpace(rc: Region[r], size: Int32): MutHeap[a, r] \ r =
+        MutHeap(rc, ref Array.new(rc, size) @ rc, ref 0 @ rc)
+
+    ///
+    /// Reserves a specific amount of space for
+    ///
+    pub def reserve(h: MutHeap[a, r], size: Int32): Unit \ r =
+        let MutHeap(rc, arr, s) = h;
+        if (size > deref s) {
+            let newArr = Array.new(rc, size);
+            Array.forEachWithIndex((x, elem) -> {
+                if (x < deref s) {
+                    Array.put(elem, x, newArr)
+                }
+                else ()
+            } , deref arr);
+            arr := newArr
+        }
+        else ()
+
+
+    ///
+    /// Returns the number of elements in a MutHeap
+    ///
+    pub def size(h: MutHeap[a, r]): Int32 \ r =
+        let MutHeap(_, _, s) = h;
+        deref s
+
+    ///
+    /// Returns whether a MutHeap is empty (has 0 elements)
+    ///
+    pub def isEmpty(h: MutHeap[a, r]): Bool \ r =
+        let MutHeap(_, _, s) = h;
+        deref s == 0
+
+    ///
+    /// Returns the top element of a MutHeap
+    ///
+    pub def peek(h: MutHeap[a, r]): Option[a] \ r =
+        let MutHeap(_, arr, s) = h;
+        if (deref s == 0) None
+        else Some(Array.get(0, deref arr))
+
+    ///
+    /// Removes and returns the top element of a MutHeap
+    ///
+    pub def pop(h: MutHeap[a, r]): Option[a] \ r with Order[a] =
+        let MutHeap(_, _, s) = h;
+        if (deref s > 0) {
+            match peek(h) {
+                case Some(elem) => {
+                    delete(h, elem);
+                    Some(elem)
+                }
+                case _          => None
+            }
+        }
+        else None
+
+    ///
+    /// Inserts an element into a MutHeap
+    ///
+    pub def insert(h: MutHeap[a, r], elem: a): Unit \ r with Order[a] = {
+        let MutHeap(_, arr, s) = h;
+        if (deref s == 0) {
+            Array.put(elem, 0, deref arr);
+            s := 1
+        }
+        else {
+            expand!(h);
+            Array.put(elem, deref s, deref arr);
+            heapifyUp(deref s, h);
+            s := deref s + 1
+        }
+    }
+
+    ///
+    /// Deletes an element from a MutHeap
+    ///
+    pub def delete(h: MutHeap[a, r], elem: a): Unit \ r with Order[a] = {
+        let MutHeap(_, arr, s) = h;
```

Comment:
Is there any way to be smarter about finding the element? E.g. if we traverse the like a tree then we can avoid branches if the node we are looking at is too big/small?

Reply:
Though I could be mistaken, I don't believe there is. This is because the heap invariant merely asserts that each node's children are smaller than it, but the children need not be ordered against each other in any specific way.

Reply:
Actually, I suppose it's possible to hasten a bit, but I think it would still be O(n)

Reply:
I don't think we should support this operation since it is expensive.